cranelift/zkasm_data/benchmarks/Design.md

```diff
@@ -0,0 +1,112 @@
+# Benchmarking infrastructure design
+
+This documents outlines the design of the infrastructure to be build for ZK WASM benchmarking. The insights we want to gain with benchmarking can be divided into the following two categories:
+
+- **Analysis** targets understanding the performance of zkASM generated by the compiler and tries to answer questions like: Which instructions are executed how often during execution and which are hot instructions? How is performance affected by changes to the compiler? How can the performance of generated zkASM be improved?
+- **Comparison** relates the performance of ZK WASM to the performance of other ZK proving systems and tries to answer questions like: Which proving system is more performant? What are the reasons for differences in performance?
+
+# Analysis
+
+This section describes the technical architecture and user interface of infrastructure and tooling related to *analysis*.
+
+## Instrumentation
+
+Instrumentation is based on logging zkASM instructions by calling Javascript from zkASM. The indirection via Javascript is required as `zkevm-proverjs` executes compiled PIL instead of zkASM. The code for zkASM instruction logging might be inserted to zkASM by enabling a compilation flag or in post processing.
+
+Subsequently raw instruction logs are processed to aggregate data and surface important information. Information to be aggregated includes the number of times an instruction was executed and the number of register allocations. Analyzing the number of register allocations is relevant as [this example](https://github.com/near/wasmtime/issues/181) shows how reducing register allocations can decrease the number of cycles.
+
+## Visualization
```

Comment:
What would be those "visualizations"? Will they be in plain text? As Images? As HTML pages? As Grafana graphs?